Reload visible model elements after recovering from the model process termination.

amarchenko opened the pull request:
#### c05340cf9093ab7e7f01d60131fc721830948353
<pre>
Reload visible model elements after recovering from the model process termination.
<a href="https://bugs.webkit.org/show_bug.cgi?id=290559">https://bugs.webkit.org/show_bug.cgi?id=290559</a>
<a href="https://rdar.apple.com/146904497">rdar://146904497</a>

Reviewed by Ada Chan.

Listening for 2 signals here. First one is closed connection to the model
process (propagated to WebCore as renderingAbruptlyStopped) that leads to
invalidation of the current ModelPlayer for HTMLModelElement. Second
signal is model element&apos;s document visibility. If the document is
visible, ModelPlayer is recreated immediately (if necessary).

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::renderingAbruptlyStopped):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::renderingAbruptlyStopped):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp:
(WebKit::ModelProcessModelPlayerManager::modelProcessConnectionDidClose):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.h:

Canonical link: <a href="https://commits.webkit.org/292859@main">https://commits.webkit.org/292859@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae14bd8743bf00c2ec63bc2cefd84b06229c6063

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97293 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/16918 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7132 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102379 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/47821 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99338 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17211 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25367 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/74127 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/31320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100296 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/13023 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/87986 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54464 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12778 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/5864 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47264 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/89969 "Build is in progress. Recent messages:") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/82821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/5948 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104399 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24370 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/83172 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/24743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84108 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82584 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20788 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/4814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/17916 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24334 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/29487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24156 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27470 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25730 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -88,6 +88,7 @@ HTMLModelElement::HTMLModelElement(const QualifiedName& tagName, Document& docum
     , m_environmentMapReadyPromise(makeUniqueRef<EnvironmentMapPromise>())
 #endif
 {
+    document.registerForVisibilityStateChangedCallbacks(*this);
```
We may want to be careful about when to register for visibility state changes, and when to update which document to register it with.

At the minimum, if the element gets moved to a new document, we should unregistered it for visibility state changes with the old doc and register with the new doc. And before this element goes away, it should be unregistered from the document.

We could do more by only registering if there's source data for the model.

`HTMLMediaElement` may be a good reference on when to register/unregister for visibility state changes.
We may want to be careful about when to register for visibility state changes, and when to update which document to register with.

At the minimum, if the element gets moved to a new document, we should unregistered it for visibility state changes with the old doc and register with the new doc. And before this element goes away, it should be unregistered from the document.

We could do more by only registering if there's source data for the model.

`HTMLMediaElement` may be a good reference on when to register/unregister for visibility state changes.

amarchenko replied:
Good point, Ada. I addressed re-registration to a new document and clean up cases.

achan00 commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -213,8 +222,14 @@ HTMLModelElement& HTMLModelElement::readyPromiseResolve()
+
     HTMLElement::didMoveToNewDocument(oldDocument, newDocument);
     sourcesChanged();
+    updateAutoplay();
```
It's not clear to me why we need this. Can you elaborate on this?

amarchenko replied:
When `HTMLModelElement` is detached from the current document `removedFromAncestor` is called. That's where Model Player is deleted. When reattach to a new document, Model Player is re-created. My intention was to not miss autoplay state but that is covered by `createModelPlayer()`.
Removed the line.

achan00 replied:
yes, `createModelPlayer()` is responsible for restoring all the states.

achan00 commented on Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp:
```diff
@@ -87,6 +87,16 @@ void ModelProcessModelPlayerManager::didReceivePlayerMessage(IPC::Connection& co
+    m_modelProcessConnection = nullptr;
+
+    for (auto& entry : m_players) {
+        if (auto player = entry.value.get())
```
Should make `auto` a `RefPtr`.

amarchenko replied:
Done.

amarchenko committed:
Reload visible model elements after recovering from the model process termination.
https://bugs.webkit.org/show_bug.cgi?id=290559
rdar://146904497

Reviewed by Ada Chan.

Listening for 2 signals here. First one is closed connection to the model
process (propagated to WebCore as renderingAbruptlyStopped) that leads to
invalidation of the current ModelPlayer for HTMLModelElement. Second
signal is model element's document visibility. If the document is
visible, ModelPlayer is recreated immediately (if necessary).

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
* Source/WebCore/Modules/model-element/HTMLModelElement.h:
* Source/WebCore/Modules/model-element/ModelPlayer.cpp:
(WebCore::ModelPlayer::renderingAbruptlyStopped):
* Source/WebCore/Modules/model-element/ModelPlayer.h:
* Source/WebCore/Modules/model-element/ModelPlayerClient.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.cpp:
(WebKit::ModelProcessModelPlayer::renderingAbruptlyStopped):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.cpp:
(WebKit::ModelProcessModelPlayerManager::modelProcessConnectionDidClose):
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.h:

Canonical link: https://commits.webkit.org/292859@main
